core/shared/src/main/scala/zio/ZSchedule.scala

```diff
@@ -662,29 +701,88 @@ object ZSchedule {
    * and then map that value with given function.
    * */
   final def doUntil[A, B](pf: PartialFunction[A, B]): Schedule[A, Option[B]] =
-    identity[A].reconsider { (a, decision) =>
-      pf.lift(a).fold(Decision.cont(decision.delay, decision.state, Option.empty[B])) { b =>
-        Decision.done(decision.delay, decision.state, Some(b))
-      }
+    new ZSchedule[Any, A, Option[B]] {
+      type State = Unit
+      val initial = ZIO.unit
+      val extract = (a: A, _: Unit) => pf.lift(a)
+      val update  = (a: A, _: Unit) => pf.lift(a).fold[IO[Unit, Unit]](ZIO.succeed(()))(_ => ZIO.fail(()))
     }
 
+  /**
+   * A schedule that will recur until the specified duration elapses. Returns
+   * the total elapsed time.
+   */
+  final def duration(duration: Duration): ZSchedule[Clock, Any, Duration] =
+    elapsed.untilOutput(_ >= duration)
+
+  /**
+   * A schedule that recurs forever without delay. Returns the elapsed time
+   * since the schedule began.
+   */
+  final val elapsed: ZSchedule[Clock, Any, Duration] =
+    ZSchedule[Clock, (Long, Long), Any, Duration](
+      clock.nanoTime.map((_, 0L)),
+      { case (_, (start, _))   => clock.nanoTime.map(currentTime => (start, currentTime - start)) },
+      { case (_, (_, elapsed)) => Duration.fromNanos(elapsed) }
+    )
+
   /**
    * A schedule that always recurs, but will wait a certain amount between
    * repetitions, given by `base * factor.pow(n)`, where `n` is the number of
    * repetitions so far. Returns the current duration between recurrences.
    */
-  final def exponential(base: Duration, factor: Double = 2.0): Schedule[Any, Duration] =
-    delayed(forever.map(i => base * math.pow(factor, i.doubleValue)))
+  final def exponential(base: Duration, factor: Double = 2.0): ZSchedule[Clock, Any, Duration] =
+    delayed(forever.map(i => base * math.pow(factor, (i + 1).doubleValue)))
 
   /**
    * A schedule that always recurs, increasing delays by summing the
    * preceding two delays (similar to the fibonacci sequence). Returns the
    * current duration between recurrences.
    */
-  final def fibonacci(one: Duration): Schedule[Any, Duration] =
-    delayed(unfold[(Duration, Duration)]((Duration.Zero, one)) {
-      case (a1, a2) => (a2, a1 + a2)
-    }.map(_._1))
+  final def fibonacci(one: Duration): ZSchedule[Clock, Any, Duration] =
```

Comment:
I really like how the `R` tells you if the schedule sleeps or not now. More information in the type.